core/src/main/clojure/xtdb/buffer_pool.clj

```diff
@@ -35,258 +48,424 @@
-                                  (when-let [^ArrowBuf cached-full-buffer (cache-get buffers k)]
-                                    (.slice cached-full-buffer start len)))]
-         (do
-           (record-cache-hit cached-buffer)
-           cached-buffer)
-
-         (let [nio-buffer (os/get-path-range root-path k start len)
-               create-arrow-buf #(util/->arrow-buf-view allocator nio-buffer)
-               [_ buf] (cache-compute buffers [k start len] create-arrow-buf)]
-           buf)))))
-
-  (putObject [_ k buf] (CompletableFuture/completedFuture (os/put-path root-path k buf)))
-  (listObjects [_this] (os/list-path root-path))
-  (listObjects [_this dir] (os/list-path root-path dir))
+    (let [cached-buffer (cache-get memory-store k)]
+      (cond
+        (nil? k)
+        (CompletableFuture/completedFuture nil)
+
+        cached-buffer
+        (do (record-cache-hit cached-buffer)
+            (CompletableFuture/completedFuture cached-buffer))
 
+        :else
+        (let [buffer-cache-path (.resolve disk-store (str k))]
+          (-> (if (util/path-exists buffer-cache-path)
+                ;; todo could this not race with eviction? e.g exists for this cond, but is evicted before we can map the file into the cache?
+                (CompletableFuture/completedFuture buffer-cache-path)
+                (CompletableFuture/failedFuture (os/obj-missing-exception k)))
+              (util/then-apply
+                (fn [path]
+                  (let [nio-buffer (util/->mmap-path path)
+                        create-arrow-buf #(util/->arrow-buf-view allocator nio-buffer)
+                        [_ buf] (cache-compute memory-store k create-arrow-buf)]
+                    buf))))))))
+  (listObjects [_]
+    (with-open [dir-stream (Files/walk disk-store (make-array FileVisitOption 0))]
+      (vec (sort (for [^Path path (iterator-seq (.iterator dir-stream))
+                       :when (Files/isRegularFile path (make-array LinkOption 0))]
+                   (str (.relativize disk-store path)))))))
+  (listObjects [_ dir]
+    (let [dir (.resolve disk-store dir)]
+      (when (Files/exists dir (make-array LinkOption 0))
+        (with-open [dir-stream (Files/newDirectoryStream dir)]
+          (vec (sort (for [^Path path dir-stream]
+                       (str (.relativize disk-store path)))))))))
+  (openChannel [this k]
+    (let [tmp-path (Files/createTempFile "bp-channel" ".arrow" (make-array FileAttribute 0))]
+      (->DiskOnlyWritableChannel this allocator memory-store disk-store (util/->file-channel tmp-path [StandardOpenOption/APPEND]) tmp-path k)))
+  (openArrowFileWriter [this k vsr]
+    (ArrowFileWriter. vsr nil (.openChannel this k)))
+  (putObject [this k buffer] (put-object this k buffer)))
+
+(deftype RemoteWritableChannel
+  [^IBufferPool bp
+   allocator
+   memory-store
+   ^Path disk-store
+   ^ObjectStore remote-store
+   ^FileChannel file-channel
+   tmp-path
+   ^String k
+   allow-file-deletion]
```

Comment:
do we use `allow-file-deletion`?

Reply:
Ha! this is a sneaky feature, it seemed reasonable to have this in non object version due to implementation sharing. Here it denotes whether or not the disk cache size should be the same as the virtual memory limit (or mappings limit). 

In my opinion, this may be a value add (which could be extended to a separate disk limit), but honestly, it snuck in as I refactored from the newer non-object base.